Added List DEFAULT partition (#13163)
pingcap/docs: partitioned-table.md

### List partitioning

Before creating a List partitioned table, make sure the following system variables are set to their default values of `ON`:

> - [`tidb_enable_list_partition`](/system-variables#tidb_enable_list_partition-new-in-v50)
> - [`tidb_enable_table_partition`](/system-variables#tidb_enable_table_partition)

Reviewer: ```suggestion
- [`tidb_enable_list_partition`](/system-variables.md#tidb_enable_list_partition-new-in-v50)
- [`tidb_enable_table_partition`](/system-variables.md#tidb_enable_table_partition)
```

You can also execute `ALTER TABLE employees DROP PARTITION pEast` to delete all related rows, but this statement also deletes the `pEast` partition from the table definition. In this situation, you must execute the `ALTER TABLE ... ADD PARTITION` statement to recover the original partitioning scheme of the table.

#### Default List partition

> Starting from v7.3.0, TiDB Cloud supports the default List partition feature, which lets you add a default List partition to a List partitioned table. The default List partition acts as a catch-all partition, where rows that do not match any other defined value sets can be placed.

Reviewer: ```suggestion
Starting from v7.3.0, TiDB supports the default List partition feature, which lets you add a default List partition to a List partitioned table. The default List partition acts as a catch-all partition, where rows that do not match any other defined value sets can be placed.
```

The default List partition feature is controlled by the [`tidb_enable_default_list_partition`](/system-variables.md#tidb_enable_default_list_partition-new-in-v730) variable, which is disabled by default because it is a TiDB extension and is not compatible with MySQL.

> When [`tidb_enable_default_list_partition`](/system-variables.md#tidb_enable_default_list_partition-new-in-v730) is set to `ON`, you can add a default partition to an existing List partition table.

Reviewer 2: Global comment: please update all `List partition table` to `List partitioned table`.
```suggestion
When [`tidb_enable_default_list_partition`](/system-variables.md#tidb_enable_default_list_partition-new-in-v730) is set to `ON`, you can add a default partition to an existing List partitioned table.
```

Author: Done.

Take the following List partition table as an example:

```sql
CREATE TABLE t (
  a INT,
  b INT
)
PARTITION BY LIST (a) (
  PARTITION p0 VALUES IN (1, 2, 3),
  PARTITION p1 VALUES IN (4, 5, 6)
);
Query OK, 0 rows affected (0.11 sec)
```

> You can add a default list partition named `pDef`to the table as follows:

Reviewer: ```suggestion
You can add a default list partition named `pDef` to the table as follows:
```

```sql
ALTER TABLE t ADD PARTITION (PARTITION pDef DEFAULT);
```

or

```sql
ALTER TABLE t ADD PARTITION (PARTITION pDef VALUES IN (DEFAULT));
```

In this way, any newly inserted values that do not match the value sets by any partitions automatically go into the default partition.

```sql
INSERT INTO t VALUES (7, 7);
Query OK, 1 row affected (0.01 sec)
```

You can also add a default partition when creating a List partitioned table. For example:

```sql
CREATE TABLE employees (
    id INT NOT NULL,
    hired DATE NOT NULL DEFAULT '1970-01-01',
    store_id INT
)
PARTITION BY LIST (store_id) (
    PARTITION pNorth VALUES IN (1, 2, 3, 4, 5),
    PARTITION pEast VALUES IN (6, 7, 8, 9, 10),
    PARTITION pWest VALUES IN (11, 12, 13, 14, 15),
    PARTITION pCentral VALUES IN (16, 17, 18, 19, 20),
    PARTITION pDefault DEFAULT
);
```

> When [`tidb_enable_default_list_partition`](/system-variables.md#tidb_enable_default_list_partition-new-in-v730) is `OFF`, List partitioning does not have a default partition to store all values that do not belong to other partitions. Therefore, all expected values of a partition expression must be included in the `PARTITION ... VALUES IN (...)` clause. If the value to be inserted in an `INSERT` statement does not match the column value set of any partition, the statement fails to execute and an error is reported. See the following example:

Reviewer: ```suggestion
When [`tidb_enable_default_list_partition`](/system-variables.md#tidb_enable_default_list_partition-new-in-v730) is `OFF`, List partition table does not support a default partition to store all values that do not belong to other partitions. Therefore, all expected values of a partition expression must be included in the `PARTITION ... VALUES IN (...)` clause. If the value to be inserted in an `INSERT` statement does not match the column value set of any partition, the statement fails to execute and an error is reported. See the following example:
```

```

or

```sql
ALTER TABLE t ADD PARTITION (PARTITION pDef VALUES IN (DEFAULT));
```

In this way, newly inserted values that do not match value sets of any partitions can automatically go into the default partition.

```sql
INSERT INTO t VALUES (7, 7);
Query OK, 1 row affected (0.01 sec)
```

You can also add a default partition when creating a List or List COLUMNS partitioned table. For example:

```sql
CREATE TABLE employees (
    id INT NOT NULL,
    hired DATE NOT NULL DEFAULT '1970-01-01',
    store_id INT
)
PARTITION BY LIST (store_id) (
    PARTITION pNorth VALUES IN (1, 2, 3, 4, 5),
    PARTITION pEast VALUES IN (6, 7, 8, 9, 10),
    PARTITION pWest VALUES IN (11, 12, 13, 14, 15),
    PARTITION pCentral VALUES IN (16, 17, 18, 19, 20),
    PARTITION pDefault DEFAULT
);
```

For a List or List COLUMNS partitioned table without a default partition, the values to be inserted using an `INSERT` statement must match value sets defined in the `PARTITION ... VALUES IN (...)` clauses of the table. If the values to be inserted do not match value sets of any partitions, the statement will fail and an error is returned, as shown in the following example:

> 

Reviewer 2: ```suggestion
```

You can also execute `ALTER TABLE employees DROP PARTITION pEast` to delete all related rows, but this statement also deletes the `pEast` partition from the table definition. In this situation, you must execute the `ALTER TABLE ... ADD PARTITION` statement to recover the original partitioning scheme of the table.

> If the new feature Default List Partition, controlled by `tidb_enable_default_list_partition` variable is not used, then unlike Range partitioning, List partitioning does not have a similar `MAXVALUE` partition to store all values that do not belong to other partitions. Instead, all expected values of the partition expression must be included in the `PARTITION ... VALUES IN (...)` clause. If the value to be inserted in an `INSERT` statement does not match the column value set of any partition, the statement fails to execute and an error is reported. See the following example:

Reviewer: @mjonss  Would you please confirm whether tidb_enable_default_list_partition is a new system variable? If yes, would you please add its description to system-variables.md in this PR?

Author: Hi, it is a new system variable. I will add it to system-variables.md

3 rows in set (0.01 sec)
```

New in v7.1 is the support of a DEFAULT list partition, which acts as a 'catch-all' partition, where all rows that does not match the defined sets of values will be placed.

> Enable this feature with `set tidb_enable_default_list_partition = ON`, it is OFF by default, since it is not supported by MySQL.

Reviewer: ```suggestion
Before using the default list partition feature, you need to set the `tidb_enable_default_list_partition` system variable to `ON`, which is `OFF` by default because it is not supported in MySQL.
```

Then add a default partition like:

```sql
ALTER TABLE t ADD PARTITION (PARTITION pDef DEFAULT);
```

or

```sql
ALTER TABLE t ADD PARTITION (PARTITION pDef VALUES IN (DEFAULT));
```

Which then will allow values that does not match the column value set by any partition to go into the Default partition.
INSERT INTO t VALUES (7, 7);
Query OK, 1 row affected (0.01 sec)
```

> One can also use it when creating the table:

Reviewer: ```suggestion
You can also add a default partition when creating the table. For example:
```
